akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -257,6 +314,98 @@ private[akka] object EventWriter {
               Behaviors.same
           }
 
+        def handleWrite(repr: PersistentRepr, replyTo: ActorRef[StatusReply[WriteAck]]): Unit = {
+          val persistenceId = repr.persistenceId
+          val sequenceNumber = repr.sequenceNr
+          val newStateForPid =
+            perPidWriteState.get(persistenceId) match {
+              case None =>
+                if (fillSequenceNumberGaps && sequenceNumber != 1L) {
+                  askMaxSeqNr(persistenceId, originalErrorDesc = None)
+                  StateForPid(
+                    waitingForReply = Map.empty,
+                    waitingForSeqNrLookup = Vector((repr, replyTo)),
+                    seqNrlookupInProgress = true,
+                    currentTransactionId = 1)
+                } else {
+                  sendToJournal(1, Vector(repr))
+                  StateForPid(
+                    Map((repr.sequenceNr, (repr, replyTo))),
+                    currentTransactionId = 1)
+                }
+
+              case Some(state) =>
+                val expectedSeqNr =
+                  if (fillSequenceNumberGaps) state.nextExpectedSeqNr
+                  else repr.sequenceNr
+
+                if (state.seqNrlookupInProgress) {
+                  context.log.trace2(
+                    "Seq nr lookup in progress for persistence id [{}], adding sequence nr [{}] to pending",
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (fillSequenceNumberGaps && state.latestSeqNr == 0L && sequenceNumber != 1L) {
+                  // we haven't looked up latest yet, and incoming is not the well known 1
```

Comment:
maybe there could be an optimization to skip this lookup if all prior sequence numbers (from 1 until waitingForWrite) are queued in waitingForReply + waitingForWrite